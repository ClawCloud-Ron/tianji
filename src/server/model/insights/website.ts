import { z } from 'zod';
import { insightsQuerySchema } from '../../utils/schema.js';
import { prisma } from '../_client.js';
import { getDateQuery, printSQL } from '../../utils/prisma.js';
import { Prisma } from '@prisma/client';
import dayjs from 'dayjs';
import { FilterInfoType, FilterInfoValue, getDateArray } from '@tianji/shared';
import { mapValues } from 'lodash-es';
import { EVENT_TYPE } from '../../utils/const.js';
import { env } from '../../utils/env.js';
import { buildCommonFilterQueryOperator } from './shared.js';

export async function insightsWebsite(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
): Promise<
  {
    date: string;
  }[]
> {
  const { time } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  const sql = buildInsightsWebsiteSql(query, context);

  if (env.isDev) {
    printSQL(sql);
  }

  const res = await prisma.$queryRaw<{ date: string | null }[]>(sql);

  return getDateArray(
    res.map((item) => {
      return {
        ...mapValues(item, (val) => Number(val)),
        date: String(item.date),
      };
    }),
    startAt,
    endAt,
    unit,
    timezone
  );
}

export function buildInsightsWebsiteSql(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
): Prisma.Sql {
  const { insightId, time, metrics, filters, groups } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  const selectQueryArr = metrics.map((item) => {
    if (item.math === 'events') {
      if (item.name === '$all_event') {
        return Prisma.sql`count(1) as "$all_event"`;
      }

      if (item.name === '$page_view') {
        return Prisma.sql`sum(case WHEN "WebsiteEvent"."eventName" is null AND "WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView} THEN 1 ELSE 0 END) as "$page_view"`;
      }

      return Prisma.sql`sum(case WHEN "WebsiteEvent"."eventName" = ${item.name} THEN 1 ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
    } else if (item.math === 'sessions') {
      if (item.name === '$all_event') {
        return Prisma.sql`count(distinct "sessionId") as "$all_event"`;
      }

      if (item.name === '$page_view') {
        return Prisma.sql`count(distinct case WHEN "WebsiteEvent"."eventName" is null AND "WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView} THEN "sessionId" ELSE null END) as "$page_view"`;
      }

      return Prisma.sql`count(distinct case WHEN "WebsiteEvent"."eventName" = ${item.name} THEN "sessionId" END) as ${Prisma.raw(`"${item.name}"`)}`;
    }
  });

  let groupSelectQueryArr: Prisma.Sql[] = [];
  if (groups.length > 0) {
    for (const g of groups) {
      if (!g.customGroups) {
        groupSelectQueryArr.push(
          Prisma.sql`${getValueField(g.type)} as "%${Prisma.raw(g.value)}"`
        );
      } else if (g.customGroups && g.customGroups.length > 0) {
        for (const cg of g.customGroups) {
          groupSelectQueryArr.push(
            Prisma.sql`${buildFilterQueryOperator(
              g.type,
              cg.filterOperator,
              cg.filterValue
            )} as "%${Prisma.raw(`${g.value}|${cg.filterOperator}|${cg.filterValue}`)}"`
          );
        }
      }
    }
  }

  let innerJoinQuery = Prisma.empty;
  if (filters.length > 0 || groups.length > 0) {
    innerJoinQuery = Prisma.sql`INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId"`;

    if (filters.length > 0) {
      innerJoinQuery = Prisma.sql`${innerJoinQuery} AND ${Prisma.join(
        filters.map((filter) =>
          buildFilterQueryOperator(filter.type, filter.operator, filter.value)
        ),
        ' AND '
      )}`;
    }

    if (groups.length > 0) {
      innerJoinQuery = Prisma.sql`${innerJoinQuery} AND ${Prisma.join(
        groups.map(
          (g) => Prisma.sql`"WebsiteEventData"."eventKey" = ${g.value}`
        ),
        ' OR '
      )}`;
    }
  }

  const whereQueryArr = [
    // website id
    Prisma.sql`"WebsiteEvent"."websiteId" = ${insightId}`,

    // date
    Prisma.sql`"WebsiteEvent"."createdAt" between ${dayjs(startAt).toISOString()}::timestamptz and ${dayjs(endAt).toISOString()}::timestamptz`,

    // event name
    Prisma.join(
      metrics.map((item) => {
        if (item.name === '$all_event') {
          return Prisma.sql`1 = 1`;
        }

        if (item.name === '$page_view') {
          return Prisma.sql`"WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView}`;
        }

        return Prisma.sql`"WebsiteEvent"."eventName" = ${item.name}`;
      }),
      ' OR ',
      '(',
      ')'
    ),
  ];

  const groupByText = Array.from({ length: groupSelectQueryArr.length + 1 })
    .map((_, i) => `${i + 1}`)
    .join(' , ');

  const sql = Prisma.sql`select
      ${getDateQuery('"WebsiteEvent"."createdAt"', unit, timezone)} date,
      ${Prisma.join([...groupSelectQueryArr, ...selectQueryArr], ' , ')}
    from "WebsiteEvent" ${innerJoinQuery}
    where ${Prisma.join(whereQueryArr, ' AND ')}
    group by ${Prisma.raw(groupByText)}`;

  return sql;
}

function getValueField(type: FilterInfoType): Prisma.Sql {
  const valueField =
    type === 'number'
      ? Prisma.sql`"WebsiteEventData"."numberValue"`
      : type === 'string'
        ? Prisma.sql`"WebsiteEventData"."stringValue"`
        : type === 'date'
          ? Prisma.sql`"WebsiteEventData"."dateValue"`
          : Prisma.sql`"WebsiteEventData"."numberValue"`;

  return valueField;
}

function buildFilterQueryOperator(
  type: FilterInfoType,
  operator: string,
  value: FilterInfoValue | null
) {
  const valueField = getValueField(type);

  return buildCommonFilterQueryOperator(type, operator, value, valueField);
}
