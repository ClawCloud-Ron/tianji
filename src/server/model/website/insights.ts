import { z } from 'zod';
import { insightsQuerySchema } from '../../utils/schema.js';
import { prisma } from '../_client.js';
import { getDateQuery } from '../../utils/prisma.js';
import { Prisma } from '@prisma/client';
import dayjs from 'dayjs';
import {
  FilterBooleanOperator,
  FilterDateOperator,
  FilterInfoType,
  FilterInfoValue,
  FilterNumberOperator,
  FilterStringOperator,
  getDateArray,
} from '@tianji/shared';
import { get, mapValues } from 'lodash-es';
import { EVENT_TYPE } from '../../utils/const.js';
import { env } from '../../utils/env.js';
import { castToDate, castToNumber, castToString } from '../../utils/cast.js';

export async function insightsWebsite(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
): Promise<
  {
    date: string;
  }[]
> {
  const { insightId, insightType, time, metrics, filters } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  const selectQueryArr = metrics.map((item) => {
    if (item.math === 'events') {
      if (item.name === '$all_event') {
        return Prisma.sql`count(1) as "$all_event"`;
      }

      if (item.name === '$page_view') {
        return Prisma.sql`sum(case WHEN "WebsiteEvent"."eventName" is null AND "WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView} THEN 1 ELSE 0 END) as "$page_view"`;
      }

      return Prisma.sql`sum(case WHEN "WebsiteEvent"."eventName" = ${item.name} THEN 1 ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
    } else if (item.math === 'sessions') {
      if (item.name === '$all_event') {
        return Prisma.sql`count(distinct "sessionId") as "$all_event"`;
      }

      if (item.name === '$page_view') {
        return Prisma.sql`count(distinct case WHEN "WebsiteEvent"."eventName" is null AND "WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView} THEN "sessionId" ELSE null END) as "$page_view"`;
      }

      return Prisma.sql`count(distinct case WHEN "WebsiteEvent"."eventName" = ${item.name} THEN "sessionId" ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
    }
  });

  let innerJoinQuery = Prisma.empty;
  if (filters.length > 0) {
    innerJoinQuery = Prisma.sql`INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId" AND ${Prisma.join(
      filters.map((filter) =>
        buildFilterQueryOperator(filter.type, filter.operator, filter.value)
      ),
      ' AND '
    )}`;
  }

  const whereQueryArr = [
    // website id
    Prisma.sql`"WebsiteEvent"."websiteId" = ${insightId}`,

    // date
    Prisma.sql`"WebsiteEvent"."createdAt" between ${dayjs(startAt).toISOString()}::timestamptz and ${dayjs(endAt).toISOString()}::timestamptz`,

    // event name
    Prisma.join(
      metrics.map((item) => {
        if (item.name === '$all_event') {
          return Prisma.sql`1 = 1`;
        }

        if (item.name === '$page_view') {
          return Prisma.sql`"WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView}`;
        }

        return Prisma.sql`"WebsiteEvent"."eventName" = ${item.name}`;
      }),
      ' OR ',
      '(',
      ')'
    ),
  ];

  const sql = Prisma.sql`select
      ${getDateQuery('"WebsiteEvent"."createdAt"', unit, timezone)} date,
      ${Prisma.join(selectQueryArr, ' , ')}
    from "WebsiteEvent" ${innerJoinQuery}
    where ${Prisma.join(whereQueryArr, ' AND ')}
    group by 1`;

  if (env.isDev) {
    console.log('insights sql:', sql.text);
  }

  const res = await prisma.$queryRaw<{ date: string | null }[]>(sql);

  return getDateArray(
    res.map((item) => {
      return {
        ...mapValues(item, (val) => Number(val)),
        date: String(item.date),
      };
    }),
    startAt,
    endAt,
    unit,
    timezone
  );
}

function buildFilterQueryOperator(
  type: FilterInfoType,
  _operator: string,
  value: FilterInfoValue | null
) {
  if (type === 'number') {
    const operator = _operator as FilterNumberOperator;

    if (operator === 'equals') {
      return Prisma.sql`"WebsiteEventData"."numberValue" = ${castToNumber(value)}`;
    }
    if (operator === 'not equals') {
      return Prisma.sql`"WebsiteEventData"."numberValue" != ${castToNumber(value)}`;
    }
    if (operator === 'in list' && Array.isArray(value)) {
      return Prisma.sql`"WebsiteEventData"."numberValue" IN (${value.join(',')})`;
    }
    if (operator === 'not in list' && Array.isArray(value)) {
      return Prisma.sql`"WebsiteEventData"."numberValue" NOT IN (${value.join(',')})`;
    }
    if (operator === 'greater than') {
      return Prisma.sql`"WebsiteEventData"."numberValue" > ${castToNumber(value)}`;
    }
    if (operator === 'greater than or equal') {
      return Prisma.sql`"WebsiteEventData"."numberValue" >= ${castToNumber(value)}`;
    }
    if (operator === 'less than') {
      return Prisma.sql`"WebsiteEventData"."numberValue" < ${castToNumber(value)}`;
    }
    if (operator === 'less than or equal') {
      return Prisma.sql`"WebsiteEventData"."numberValue" <= ${castToNumber(value)}`;
    }

    if (operator === 'between') {
      return Prisma.sql`"WebsiteEventData"."numberValue" BETWEEN ${castToNumber(get(value, '0'))} AND ${castToNumber(get(value, '1'))}`;
    }
  } else if (type === 'string') {
    const operator = _operator as FilterStringOperator;

    if (operator === 'equals') {
      return Prisma.sql`"WebsiteEventData"."stringValue" = ${castToString(value)}`;
    }
    if (operator === 'not equals') {
      return Prisma.sql`"WebsiteEventData"."stringValue" != ${castToString(value)}`;
    }
    if (operator === 'contains') {
      return Prisma.sql`"WebsiteEventData"."stringValue" LIKE "%${castToString(value)}%"`;
    }
    if (operator === 'not contains') {
      return Prisma.sql`"WebsiteEventData"."stringValue" NOT LIKE "%${castToString(value)}%"`;
    }
    if (operator === 'in list' && Array.isArray(value)) {
      return Prisma.sql`"WebsiteEventData"."stringValue" IN (${value.join(',')})`;
    }
    if (operator === 'not in list' && Array.isArray(value)) {
      return Prisma.sql`"WebsiteEventData"."stringValue" NOT IN (${value.join(',')})`;
    }
  } else if (type === 'boolean') {
    const operator = _operator as FilterBooleanOperator;

    if (operator === 'equals') {
      return Prisma.sql`"WebsiteEventData"."numberValue" = ${castToNumber(value)}`;
    }
    if (operator === 'not equals') {
      return Prisma.sql`"WebsiteEventData"."numberValue" != ${castToNumber(value)}`;
    }
  } else if (type === 'date') {
    const operator = _operator as FilterDateOperator;

    if (operator === 'between') {
      return Prisma.sql`"WebsiteEventData"."dateValue" BETWEEN ${castToDate(get(value, '0'))} AND ${castToDate(get(value, '1'))}`;
    }
    if (operator === 'in day') {
      return Prisma.sql`"WebsiteEventData"."dateValue" = DATE(${castToDate(value)})`;
    }
  }

  return Prisma.sql`1 = 1`;
}
