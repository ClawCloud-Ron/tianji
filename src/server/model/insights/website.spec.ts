import { describe, expect, test } from 'vitest';
import { unwrapSQL } from '../../utils/prisma.js';
import { WebsiteInsightsSqlBuilder } from './website.js';

describe('WebsiteInsightsSqlBuilder', () => {
  const insightId = 'cly5yay7a001v5tp6xdkzmygh';
  const insightType = 'website';

  test('groups', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'number',
            type: 'number',
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });

  test('groups with custom bucket', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'number',
            type: 'number',
            customGroups: [
              {
                filterOperator: 'not equals',
                filterValue: 1,
              },
            ],
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });
});
