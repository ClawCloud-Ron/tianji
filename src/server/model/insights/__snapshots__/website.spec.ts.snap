// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WebsiteInsightsSqlBuilder > groups > sql 1`] = `
"select
      to_char(date_trunc('day', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date,
      "WebsiteEventData"."numberValue" as "%number" , count(1) as "$all_event"
    from "WebsiteEvent" INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId" AND "WebsiteEventData"."eventKey" = 'number'
    where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND (1 = 1)
    group by 1 , 2"
`;

exports[`WebsiteInsightsSqlBuilder > groups with custom bucket > sql 1`] = `
"select
      to_char(date_trunc('day', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date,
      "WebsiteEventData"."numberValue" != '1' as "%number|not equals|1" , count(1) as "$all_event"
    from "WebsiteEvent" INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId" AND "WebsiteEventData"."eventKey" = 'number'
    where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND (1 = 1)
    group by 1 , 2"
`;
